bitcoin/src/address/mod.rs

```diff
@@ -186,9 +132,161 @@ impl NetworkValidation for NetworkUnchecked {
+            P2pkh { hash, prefix } => {
+                let mut prefixed = [0; 21];
+                prefixed[0] = prefix.to_u8();
+                prefixed[1..].copy_from_slice(&hash[..]);
+                base58::encode_check_to_fmt(fmt, &prefixed[..])
+            }
+            P2sh { hash, prefix } => {
+                let mut prefixed = [0; 21];
+                prefixed[0] = prefix.to_u8();
+                prefixed[1..].copy_from_slice(&hash[..]);
+                base58::encode_check_to_fmt(fmt, &prefixed[..])
+            }
+            Segwit { program, hrp } => {
+                let hrp = hrp.to_hrp();
+                let version = program.version().to_fe();
+                let program = program.program().as_ref();
+
+                if fmt.alternate() {
+                    bech32::segwit::encode_upper_to_fmt_unchecked(fmt, &hrp, version, program)
+                } else {
+                    bech32::segwit::encode_lower_to_fmt_unchecked(fmt, &hrp, version, program)
+                }
+            }
+        }
+    }
+}
+
+/// Prefix byte used for legacy P2PKH addresses.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum LegacyP2pkhPrefix {
+    /// Prefix used for legacy addresses on the main Bitcoin network.
+    Mainnet,
+    /// Prefix used for legacy addresses on all other test networks (testnet, signet, regtest).
+    AllTestnets,
+}
+
+impl LegacyP2pkhPrefix {
+    /// Creates a legacy prefix from the associated `network`.
+    fn from_network(network: Network) -> Self {
+        use Network::*;
+        match network {
+            Bitcoin => Self::Mainnet,
+            Signet | Testnet | Regtest => Self::AllTestnets,
+        }
+    }
+
+    /// Converts this prefix enum to the respective byte value.
+    fn to_u8(&self) -> u8 {
+        match *self {
+            Self::Mainnet => PUBKEY_ADDRESS_PREFIX_MAIN,
+            Self::AllTestnets => PUBKEY_ADDRESS_PREFIX_TEST,
+        }
+    }
+}
+
+/// Prefix byte used for legacy P2SH addresses.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum LegacyP2shPrefix {
+    /// Prefix used for legacy addresses on the main Bitcoin network.
+    Mainnet,
+    /// Prefix used for legacy addresses on all other test networks (testnet, signet, regtest).
+    AllTestnets,
+}
+
+impl LegacyP2shPrefix {
+    /// Creates a legacy prefix from the associated `network`.
+    fn from_network(network: Network) -> Self {
+        use Network::*;
+        match network {
+            Bitcoin => Self::Mainnet,
+            Signet | Testnet | Regtest => Self::AllTestnets,
+        }
+    }
+
+    /// Converts this prefix enum to the respective byte value.
+    fn to_u8(&self) -> u8 {
+        match *self {
+            Self::Mainnet => SCRIPT_ADDRESS_PREFIX_MAIN,
+            Self::AllTestnets => SCRIPT_ADDRESS_PREFIX_TEST,
+        }
+    }
+}
+
+/// Known bech32 human-readable parts.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum KnownHrp {
```

Comment:
Ha, a nice little lesson to think more when I code, thanks. 

I added a leading patch to remove all the `PartialOrd` and `Ord` derives in the `address/mod.rs` file that were unneeded. Then fixed the ones that were in this change set.

Reply:
Made the `KnownHpr` type private, and added you suggested attributes.

Reply:
6 hours later and I worked out why we throw `Ord` on things that do not have a natural order - its so we can put them in maps. So regtest may not be `<` mainnet but its a valid usecase as a map key, right? Especially if combined with something else `BTreeMap<(Network, Foo), V>`?

Reply:
Why not use `HashMap` or `LinkedHashMap` or other maps?

But sure, it may be handy for some `no_std` applications to just use `BTreeMap`. I just don't want to be too eager implementing traits.